crates/roc_host/src/lib.rs

```diff
@@ -614,57 +634,73 @@ pub extern "C" fn roc_fx_fileReadBytes(roc_path: &RocList<u8>) -> RocResult<RocL
+            match alloc_result {
+                Ok(out) => RocResult::ok(out),
+                Err(err) => RocResult::err(toRocReadError(err)),
+            }
+        }
         Err(err) => RocResult::err(toRocReadError(err)),
     }
 }
 
 #[no_mangle]
-pub extern "C" fn roc_fx_fileReadLine(readerIndex: u64) -> RocResult<RocList<u8>, RocStr> {
-    READERS.with(|reader_thread_local| {
-        let readers = reader_thread_local.borrow_mut();
-
-        match readers.get(&readerIndex) {
-            // return an empty list when no reader was found
-            None => RocResult::ok(RocList::empty()),
-            Some(reader_ref_cell) => {
-                let mut string_buffer = String::new();
-
-                let mut file_buf_reader = (**reader_ref_cell).borrow_mut();
-
-                match file_buf_reader.read_line(&mut string_buffer) {
-                    Ok(..) => {
-                        // Note: this returns an empty list when no bytes were read, e.g. End Of File
-                        RocResult::ok(RocList::from(string_buffer.as_bytes()))
-                    }
-                    Err(err) => RocResult::err(err.to_string().as_str().into()),
-                }
-            }
+pub extern "C" fn roc_fx_fileReadLine(data: RocBox<()>) -> RocResult<RocList<u8>, RocStr> {
+    let buf_reader: &mut BufReader<File> = ThreadSafeRefcountedResourceHeap::box_to_resource(data);
+
+    let mut buffer = RocList::empty();
+    match read_until(buf_reader, '\n' as u8, &mut buffer) {
+        Ok(..) => {
+            // Note: this returns an empty list when no bytes were read, e.g. End Of File
+            RocResult::ok(buffer)
         }
-    })
+        Err(err) => RocResult::err(err.to_string().as_str().into()),
+    }
 }
 
-#[no_mangle]
-pub extern "C" fn roc_fx_closeFile(readerIndex: u64) {
-    READERS.with(|reader_thread_local| {
-        let mut readers = reader_thread_local.borrow_mut();
-
-        // Rust will close the file after this removal
-        readers.remove(&readerIndex);
-    })
+fn read_until<R: BufRead + ?Sized>(
+    r: &mut R,
+    delim: u8,
+    buf: &mut RocList<u8>,
+) -> io::Result<usize> {
+    let mut read = 0;
+    loop {
+        let (done, used) = {
+            let available = match r.fill_buf() {
+                Ok(n) => n,
+                Err(ref e) if matches!(e.kind(), ErrorKind::Interrupted) => continue,
+                Err(e) => return Err(e),
+            };
+            match memchr::memchr(delim, available) {
+                Some(i) => {
+                    buf.extend_from_slice(&available[..=i]);
+                    (true, i + 1)
+                }
+                None => {
+                    buf.extend_from_slice(available);
+                    (false, available.len())
+                }
+            }
+        };
+        r.consume(used);
+        read += used;
+        if done || used == 0 {
+            return Ok(read);
+        }
+    }
```

Comment:
Note: this is just a direct copy of the standard library function but using a roc list.

Reply:
Does this mean we can get rid of the comments elsewhere about how we want our own `read_to_end` that directly fills a list?

Reply:
`read_to_end` is different from `read_util` and it was way more complex, so I didn't port it for this PR. All the TODOs are correct.